prow/plank/reconciler.go

```diff
@@ -0,0 +1,645 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	prowv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pjutil"
+	"k8s.io/test-infra/prow/pod-utils/decorate"
+)
+
+const controllerName = "plank"
+
+func Add(
+	mgr controllerruntime.Manager,
+	buildMgrs map[string]controllerruntime.Manager,
+	cfg config.Getter,
+	totURL string,
+	additionalSelector string,
+) error {
+	return add(mgr, buildMgrs, cfg, totURL, additionalSelector, nil, nil, 10)
+}
+
+func add(
+	mgr controllerruntime.Manager,
+	buildMgrs map[string]controllerruntime.Manager,
+	cfg config.Getter,
+	totURL string,
+	additionalSelector string,
+	overwriteReconcile func(reconcile.Request) (reconcile.Result, error),
+	predicateCallack func(bool),
+	numWorkers int,
+) error {
+	predicate, err := predicates(additionalSelector, predicateCallack)
+	if err != nil {
+		return fmt.Errorf("failed to construct predicate: %w", err)
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(&prowv1.ProwJob{}, prowJobIndexName, prowJobIndexer(cfg().ProwJobNamespace)); err != nil {
+		return fmt.Errorf("failed to add indexer: %w", err)
+	}
+
+	blder := controllerruntime.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		For(&prowv1.ProwJob{}).
+		WithEventFilter(predicate).
+		WithOptions(controller.Options{MaxConcurrentReconciles: numWorkers})
+
+	r := &reconciler{
+		ctx:                context.Background(),
+		pjClient:           mgr.GetClient(),
+		buildClients:       map[string]ctrlruntimeclient.Client{},
+		overwriteReconcile: overwriteReconcile,
+		log:                logrus.NewEntry(logrus.StandardLogger()),
+		config:             cfg,
+		totURL:             totURL,
+		clock:              clock.RealClock{},
+	}
+	for buildCluster, buildClusterMgr := range buildMgrs {
+		blder = blder.Watches(
+			source.NewKindWithCache(&corev1.Pod{}, buildClusterMgr.GetCache()),
+			podEventRequestMapper(cfg().ProwJobNamespace))
+		r.buildClients[buildCluster] = buildClusterMgr.GetClient()
+	}
+
+	if err := blder.Complete(r); err != nil {
+		return fmt.Errorf("failed to build controller: %w", err)
+	}
+
+	if err := mgr.Add(manager.RunnableFunc(r.syncMetrics)); err != nil {
+		return fmt.Errorf("failed to add metrics runnable to manager: %w", err)
+	}
+
+	return nil
+}
+
+type reconciler struct {
+	ctx                context.Context
+	pjClient           ctrlruntimeclient.Client
+	buildClients       map[string]ctrlruntimeclient.Client
+	overwriteReconcile func(reconcile.Request) (reconcile.Result, error)
+	log                *logrus.Entry
+	config             config.Getter
+	totURL             string
+	clock              clock.Clock
+}
+
+func (r *reconciler) syncMetrics(stop <-chan struct{}) error {
+	for {
+		select {
+		case <-stop:
+			return nil
+		case <-time.NewTicker(30 * time.Second).C:
```

Comment:
Isn't this the form that creates new tickers and leaks them?

Reply:
Updated it